File: leetcode/659.py
class Solution:
    def isPossible(self, nums: List[int]) -> bool:
        tail_to_lens: Dict[int, Dict[int, int]] = {}
        for num in nums:
            if num not in tail_to_lens:
                tail_to_lens[num] = {}
            if num-1 not in tail_to_lens or len(tail_to_lens[num-1]) == 0:
                if 1 not in tail_to_lens[num]:
                    tail_to_lens[num][1] = 0
                tail_to_lens[num][1] += 1
            else:
                for i in [1, 2, 3]:
                    if i in tail_to_lens[num-1]:
                        tail_to_lens[num-1][i] -= 1
                        if tail_to_lens[num-1][i] == 0:
                            del tail_to_lens[num-1][i]
                        if i == 3:
                            if i not in tail_to_lens[num]:
                                tail_to_lens[num][i] = 0
                            tail_to_lens[num][i] += 1
                        else:
                            if i+1 not in tail_to_lens[num]:
                                tail_to_lens[num][i+1] = 0
                            tail_to_lens[num][i+1] += 1
                        break
        for tail in tail_to_lens:
            for i in [1, 2]:
                if i in tail_to_lens[tail]:
                    return False
        return True

File: leetcode/659.cpp
class Solution
{
public:
    bool isPossible(vector<int> &nums)
    {
        unordered_map<int, priority_queue<int> > tail_to_neg_lens;
        for (auto num : nums)
        {
            auto prev_len = 0;
            if (tail_to_neg_lens.find(num - 1) != tail_to_neg_lens.end())
            {
                auto &prev_pq = tail_to_neg_lens[num - 1];
                if (prev_pq.size())
                {
                    prev_len = prev_pq.top();
                    prev_pq.pop();
                }
            }
            tail_to_neg_lens[num].push(prev_len - 1);
        }
        for (auto p : tail_to_neg_lens)
        {
            auto pq = p.second;
            while (pq.size())
            {
                if (abs(pq.top()) < 3)
                    return false;
                pq.pop();
            }
        }
        return true;
    }
};